File: backend-laravel/app/Models/CicloLectivo.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\HasMany;

/**
 * Ciclo lectivo (tabla `ciclos_lectivos`): el año. Es el eje al que
 * cuelga todo lo transaccional del sistema (narrativa, "Traslado de
 * datos entre ciclos lectivos" — cursos, grupos, inscripciones y
 * asistencia se asocian a un ciclo; roles, permisos, especialidades y
 * parámetros de cálculo NO, porque son datos permanentes de la
 * institución).
 *
 * Sin SoftDeletes a propósito: un ciclo lectivo no se "borra", se cierra
 * (`estado = cerrado`) y queda archivado de solo lectura — ver
 * `Documentacion_Base_de_Datos.pdf`, sección 6.2.
 */
class CicloLectivo extends Model
{
    protected $table = 'ciclos_lectivos';

    protected $primaryKey = 'id_ciclo_lectivo';

    protected $fillable = [
        'anio',
        'fecha_inicio',
        'fecha_fin',
        'estado',
        'fecha_cierre',
    ];

    protected function casts(): array
    {
        return [
            'fecha_inicio' => 'date',
            'fecha_fin' => 'date',
            'fecha_cierre' => 'datetime',
        ];
    }

    public function cursos(): HasMany
    {
        return $this->hasMany(Curso::class, 'ciclo_lectivo_id', 'id_ciclo_lectivo');
    }

    public function gruposTaller(): HasMany
    {
        return $this->hasMany(GrupoTaller::class, 'ciclo_lectivo_id', 'id_ciclo_lectivo');
    }

    public function gruposEdFisica(): HasMany
    {
        return $this->hasMany(GrupoEdFisica::class, 'ciclo_lectivo_id', 'id_ciclo_lectivo');
    }

    public function inscripciones(): HasMany
    {
        return $this->hasMany(Inscripcion::class, 'ciclo_lectivo_id', 'id_ciclo_lectivo');
    }
}

File: app_flutter/lib/models/ciclo_lectivo.dart
/// Espejo de lo que devuelve `CiclosLectivosController` (backend-laravel)
/// — el ciclo lectivo (el año) del que cuelga todo lo transaccional del
/// sistema: cursos, grupos, inscripciones y asistencia. Ver
/// `App\Models\CicloLectivo` en el backend.
class CicloLectivo {
  const CicloLectivo({
    required this.id,
    required this.anio,
    required this.fechaInicio,
    required this.fechaFin,
    required this.estado,
    required this.fechaCierre,
  });

  factory CicloLectivo.fromJson(Map<String, dynamic> json) {
    return CicloLectivo(
      id: json['id_ciclo_lectivo'] as int,
      anio: json['anio'] as int,
      fechaInicio: json['fecha_inicio'] as String?,
      fechaFin: json['fecha_fin'] as String?,
      estado: json['estado'] as String,
      fechaCierre: json['fecha_cierre'] as String?,
    );
  }

  final int id;
  final int anio;
  final String? fechaInicio;
  final String? fechaFin;

  /// 'abierto' | 'cerrado'.
  final String estado;
  final String? fechaCierre;

  bool get abierto => estado == 'abierto';
}
